src/Http/Routing/src/EndpointRoutingMiddleware.cs

```diff
@@ -56,68 +56,75 @@ internal sealed class EndpointRoutingMiddleware
 
         public Task Invoke(HttpContext httpContext)
         {
-            var feature = new EndpointSelectorContext();
+            var feature = new EndpointSelectorContext(httpContext);
+
+            // Store the previous endpoint before matching starts, this lets us restore whatever endpoint was
+            // set at the end of matching if nothing matched
+            var previousEndpoint = feature.Endpoint;
```

Comment:
Is it worth a discussion about whether the current behaviour of the middleware is even desirable? For instance if the routing middleware no-op'ed when there was an endpoint does taht make more sense? 

I don't feel like we ever really answered these questions because we didn't have good scenarios in mind either way.

Reply:
I was actually thinking that was a better behavior.

Reply:
I'm going to implement this as it simplifies alot

Reply:
I'll be annoying and ask whether it should be an option

Reply:
No.

Reply:
You're not being annoying - but it shouldn't be an option. This should work (and be documented) behaving in a single way.

Reply:
We can make it an option when the masses complain (I mean @sebastienros )

Reply:
Sure. I'm fine with it not being an option straight away, if ever.